File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/repositories/infra/IUnitOfWork.java
package com.oddly.ddd.infrastructure.repositories.infra;

import java.util.concurrent.CompletableFuture;

/**
 * Unit of Work pattern interface for managing database transactions.
 * Transactions are managed automatically by UnitOfWork middleware.
 * Services should NOT manually manage transactions.
 * 
 * Lifecycle:
 * 1. Middleware calls beginTransactionAsync() before controller
 * 2. Controller and service execute business logic
 * 3. Middleware calls commitAsync() on success or rollbackAsync() on error
 */
public interface IUnitOfWork {
    /**
     * Begins a new database transaction.
     * Called automatically by UnitOfWork middleware.
     */
    CompletableFuture<Void> beginTransactionAsync();

    /**
     * Commits the current transaction.
     * Called automatically by middleware on successful response (status < 400).
     */
    CompletableFuture<Void> commitAsync();

    /**
     * Rolls back the current transaction.
     * Called automatically by middleware on error (status >= 400 or exception).
     */
    CompletableFuture<Void> rollbackAsync();

    /**
     * Saves all pending changes to the database.
     * Used within a transaction to persist changes.
     */
    CompletableFuture<Void> saveChangesAsync();
}

File: infrastructures/ddd/csharp/src/Infrastructure/Repositories/IUnitOfWork.cs
using System;
using System.Threading.Tasks;

namespace OddlyDdd.Infrastructure.Repositories
{
    /// <summary>
    /// Unit of Work pattern for managing transactions.
    /// The UnitOfWork middleware handles ALL transaction management.
    /// Services should NOT manually manage transactions.
    /// </summary>
    public interface IUnitOfWork : IDisposable
    {
        /// <summary>
        /// Begins a new transaction.
        /// Called by UnitOfWork middleware at the start of a request.
        /// </summary>
        Task BeginTransactionAsync();

        /// <summary>
        /// Commits the current transaction.
        /// Called by UnitOfWork middleware when the response status code is &lt; 400.
        /// </summary>
        Task CommitAsync();

        /// <summary>
        /// Rolls back the current transaction.
        /// Called by UnitOfWork middleware when an error occurs or status code is &gt;= 400.
        /// </summary>
        Task RollbackAsync();

        /// <summary>
        /// Saves changes to the database within the current transaction.
        /// </summary>
        Task SaveChangesAsync();
    }
}
